src/KOKKOS/fix_rigid_kokkos.cpp

```diff
@@ -0,0 +1,1274 @@
+                x_arr_type moments_ibody, double dtq)
+{
+  // full update from dq/dt = 1/2 w q
+
+	double wq[4];
+  MathExtraKokkos::vecquat(w_ibody,q_ibody,wq);
+
+  double qfull[4];
+  qfull[0] = q_ibody[0] + dtq * wq[0];
+  qfull[1] = q_ibody[1] + dtq * wq[1];
+  qfull[2] = q_ibody[2] + dtq * wq[2];
+  qfull[3] = q_ibody[3] + dtq * wq[3];
+  MathExtraKokkos::qnormalize(qfull);
+
+  // 1st half update from dq/dt = 1/2 w q
+
+  double qhalf[4];
+  qhalf[0] = q_ibody[0] + 0.5*dtq * wq[0];
+  qhalf[1] = q_ibody[1] + 0.5*dtq * wq[1];
+  qhalf[2] = q_ibody[2] + 0.5*dtq * wq[2];
+  qhalf[3] = q_ibody[3] + 0.5*dtq * wq[3];
+  MathExtraKokkos::qnormalize(qhalf);
+  
+
+  // re-compute omega at 1/2 step from m at 1/2 step and q at 1/2 step
+  // recompute wq
+  
+  MathExtraKokkos::mq_to_omega(m_ibody,qhalf,moments_ibody,w_ibody);
+  MathExtraKokkos::vecquat(w_ibody,qhalf,wq);
+
+  // 2nd half update from dq/dt = 1/2 w q
+
+  qhalf[0] += 0.5*dtq * wq[0];
+  qhalf[1] += 0.5*dtq * wq[1];
+  qhalf[2] += 0.5*dtq * wq[2];
+  qhalf[3] += 0.5*dtq * wq[3];
+  MathExtraKokkos::qnormalize(qhalf);
+  
+  // corrected Richardson update
+  
+  q_ibody[0] = 2.0*qhalf[0] - qfull[0];
+  q_ibody[1] = 2.0*qhalf[1] - qfull[1];
+  q_ibody[2] = 2.0*qhalf[2] - qfull[2];
+  q_ibody[3] = 2.0*qhalf[3] - qfull[3];
+   MathExtraKokkos::qnormalize(q_ibody);
+}
+  
+
+} // MathExtraKokkos
+
+
+
+template <class DeviceType>
+FixRigidKokkos<DeviceType>::FixRigidKokkos(LAMMPS *lmp, int narg, char **arg) :
+  FixRigid(lmp, narg, arg)
+{
+  kokkosable = 1;
+  atomKK = (AtomKokkos *) atom;
+  execution_space = ExecutionSpaceFromDevice<DeviceType>::space;
+
+  datamask_read   = (X_MASK | V_MASK | F_MASK | MASK_MASK | RMASS_MASK | TYPE_MASK
+                     | OMEGA_MASK | ANGMOM_MASK | TORQUE_MASK);
+  datamask_modify = (X_MASK | V_MASK | OMEGA_MASK | ANGMOM_MASK);
+
+  // Only some arrays are initialized in the FixRigid constructor that we need.
+  // The following are all either
+  //   a) set after that, so we need not worry about preserving the data, or
+  //   b) Set to all zeros which is also what Kokkos does by default.
+
+  // These are all zero-initialized in FixRigid::FixRigid(...) and
+  // later initialized:
+  memoryKK->create_kokkos(k_xcm,xcm,nbody,3,"rigid/kk:xcm");
+  memoryKK->create_kokkos(k_vcm,vcm,nbody,3,"rigid/kk:vcm");
+  memoryKK->create_kokkos(k_fcm,fcm,nbody,3,"rigid/kk:fcm");
+
+  memoryKK->create_kokkos(k_torque,torque,nbody,3,"rigid/kk:torque");
+
+
+  // These are initialized in the base c-tor but not to 0:
+  memoryKK->create_kokkos(k_fflag,fflag,nbody,3,"rigid/kk:fflag");
+  memoryKK->create_kokkos(k_tflag,tflag,nbody,3,"rigid/kk:tflag");
+
+  
+  // The initialization for these two is simple enough:
+  for (int i = 0; i < nbody; i++) {
+    k_fflag.d_view(i,0) = k_fflag.d_view(i,1) = k_fflag.d_view(i,2) = 1.0;
+    k_tflag.d_view(i,0) = k_tflag.d_view(i,1) = k_tflag.d_view(i,2) = 1.0;
```

Comment:
This needs to use the host view or the loop needs to be changed to a `Kokkos::parallel_for`, otherwise it will fail on GPUs because you are accessing device memory from the host. You could also use `Kokkos::deep_copy(k_fflag.d_view,1.0);`.

Reply:
This got removed so is no longer required.